### ISSUE TITLE
Scenario: Ability to connect to a remote azure resource from development

### ISSUE THREAD

**Issue by davidfowl** [dotnet/aspire#1758]
Today we ship [Aspire.Hosting.Azure.Provisioning](https://github.com/dotnet/aspire/tree/main/src/Aspire.Hosting.Azure.Provisioning), which is a package that provisions azure resources for development (with the appropriate role assignments) and saves the appropriate connection information in the apphost's configuration (user secrets).

We'd like to move this logic to be fully azd based. The idea is that running azd provision would create the manifest, provision the resources defined the apphost and write the connection information to user secrets (https://github.com/Azure/azure-dev/issues/3202).

This would better align the "dev against live resource" story and the "deploy those resources story".

Related to https://github.com/dotnet/aspire/issues/894

**Comment by bradygaster** [comment]
With [AugmentR](https://github.com/bradygaster/augmentr) and a few other AZD-deployed apps I've encountered a pattern, especially when dealing with Azure Service Bus or Event Grid, or when coding against Azure OpenAI. The pattern of finding a sample through having that sample running in Azure looks something like this: 

```mermaid
graph TD;
    Clone-->Provision;
    Provision-->Run;
    Run-->Deploy;
```

I think this is the paradigm @davidfowl is proposing, and it adheres nice to some of the work going on in the Azure SDK and Dev Experience teams. With AugmentR, which has recently been updated to use AZD Core since we're not quite ready with some of the "automagic" `azd up` features yet, a dev needs to have an Azure OpenAI instance provisioned inclusive of two specific model deployments - GPT and text embedding - to facilitate all the RAG features. You need to **provision** these resources before you can **F5/run/debug** the code. 

But it's more complex than just provisioning; once the resources exist, they need to be stored - as secrets - in `user-secrets` for use during development. 

So, to accomplish all of this:

1. Provision the resources
2. Get all the configuration values
3. Create secrets from those using `dotnet user-secrets`

We did this using a post-provision hook and a `ps1` file:

First, the [post-provision hook](https://github.com/bradygaster/AugmentR/blob/main/azure.yaml#L11):

```yaml
# yaml-language-server: $schema=https://raw.githubusercontent.com/Azure/azure-dev/main/schemas/v1.0/azure.yaml.json

name: AugmentR
metadata:
  template: AugmentR@0.0.1-beta

hooks:
  postprovision:
    windows:
      shell: pwsh
      run: ./postprovision.ps1
      interactive: true
      continueOnError: true
``` 

Then, the home-grown `ps1` [function ](https://github.com/bradygaster/AugmentR/blob/main/postprovision.ps1)that uses each of `azd's` environment variables as secrets: 

```powershell
function Set-DotnetUserSecrets {
    param ($path, $lines)
    Push-Location
    cd $path
    dotnet user-secrets clear
    foreach ($line in $lines) {
        $name, $value = $line -split '='
        $value = $value -replace '"', ''
        $name = $name -replace '__', ':'
        if ($value -ne '') {
            dotnet user-secrets set $name $value
        }
    }
    Pop-Location
}

$lines = (azd env get-values) -split "`n"
Set-DotnetUserSecrets -path ".\Backend" -lines $lines
Set-DotnetUserSecrets -path ".\Frontend" -lines $lines
```

@davidfowl do you feel like this approach - not this solution, but, the provision-with-azd-prior-to-debug scenario - is sufficient as an experience? 

(Note: I need to update the SK code in AugmentR so it isn't building atm, but, it serves the purpose of demonstrating a "success state" of provision-with-azd-prior-to-debug that felt relevant here)

**Comment by davidfowl** [comment]
> @davidfowl do you feel like this approach - not this solution, but, the provision-with-azd-prior-to-debug scenario - is sufficient as an experience?

This is the entire feature being proposed so I love it! 😄 

Some other problems with azd provision is that today it provisions more than needed for development scenarios (container registray, container apps env, etc etc). I don't know if this needs to be a provision lite or something else.

**Comment by bradygaster** [comment]
> Some other problems with azd provision is that today it provisions more than needed for development scenarios (container registray, container apps env, etc etc). I don't know if this needs to be a provision lite or something else.

Given a file-new-yes-to-Redis Aspire scenario being published to ACA, how could we mitigate an ACR instance? Gotta have a registry, right? On that same front, what would we "not need" that AZD lays down? My 2C is we don't need a provision lite; I love the way AZD does a whole-environment initialization sort of thing. 

Maybe I set up too many demos and customer studies. :)

**Comment by davidfowl** [comment]
Right this is broken today with this idea since it would spin up the wrong kind of redis instance. I think in this mode (dependency only mode), we would only spin up azure dependencies. Spinning up containers remotely doesn't work.

**Comment by bradygaster** [comment]
Yeah i think this is where @ellismg's idea of `profiles` seemed pretty nifty. I still prefer the explicitness of AZD environments, but, there is some merit to the questions like "you say redis but what do you reeeeeeally mean when you say r-e-d-i-s for this particular environment?"

**Comment by davidfowl** [comment]
The last thing we decided was that the information about what azure resources get deployed go into the app host. That's what [this scenario is about](https://github.com/dotnet/aspire/issues/1757).

**Comment by seesharprun** [comment]
> ```mermaid
> graph TD;
>     Clone-->Provision;
>     Provision-->Run;
>     Run-->Deploy;
> ```

@bradygaster, I'm doing something very similar for a .NET Orleans quickstart where the workflow is:

1. Init the AZD template for url shortener
1. `azd up` to provision and deploy. The URL shortener uses in-memory persistence by default
1. Change an AZD variable with `azd env set` to select Azure Table Storage or Azure Cosmos DB for NoSQL to deploy
1. Deploy with `azd provision` or `azd up`.
1. Pull NuGet packages and implement alternative persistence
1. `azd deploy` or `azd up` the new code.

It's a great flow to obtain the sample, iterate changes, and push them out.

I sometimes work on build hands-on labs that feature Azure OpenAI and the idea of "dev against the real service from the start" is real with every lab starting with an Azure OpenAI deployment (instead of ending with one like with other services). It's also helpful to start with a working fully deployed application from the beginning instead of trying to piecemeal it so you can troubleshoot if your iterative changes go sideways.
